extensions/parquet/table/src/main/java/io/deephaven/parquet/table/util/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,320 @@
+        final int currFragmentIndex = fragmentIndexForByteNumber(localPosition);
+        CompletableFuture<ByteBuffer> currFragmentFuture = context.getCachedFuture(currFragmentIndex);
+        if (currFragmentFuture == null) {
+            currFragmentFuture = computeFragmentFuture(currFragmentIndex);
+            context.setFragmentContext(currFragmentIndex, currFragmentFuture);
+        }
+
+        // Send async requests for read-ahead buffers and store them in the cache
+        final int numFragmentsToLoad = Math.min(readAheadCount, numFragmentsInObject - currFragmentIndex - 1);
+        for (int i = 0; i < numFragmentsToLoad; i++) {
+            final int readAheadFragmentIndex = i + currFragmentIndex + 1;
+            final CompletableFuture<ByteBuffer> readAheadFragmentFuture = context.getCachedFuture(readAheadFragmentIndex);
+            if (readAheadFragmentFuture == null) {
+                context.setFragmentContext(readAheadFragmentIndex, computeFragmentFuture(readAheadFragmentIndex));
+            }
+        }
+
+        // Wait till the current fragment is fetched
+        final ByteBuffer currentFragment;
+        try {
+            currentFragment = currFragmentFuture.get(readTimeout.toNanos(), TimeUnit.NANOSECONDS);
+        }  catch (final InterruptedException | ExecutionException | TimeoutException e) {
+            final String operation = "fetching fragment " + currFragmentIndex + " for file " + key + " in S3 bucket " + bucket;
+            handleS3Exception(e, operation);
+            throw new UncheckedDeephavenException("Exception caught while " + operation, e);
+        }
+
+        // Copy the bytes from fragment from the offset up to the min of remaining fragment and destination bytes.
+        // Therefore, the number of bytes read by this method can be less than the number of bytes remaining in the
+        // destination buffer.
+        final int fragmentOffset = (int) (localPosition - (currFragmentIndex * fragmentSize));
+        currentFragment.position(fragmentOffset);
+        final int limit = Math.min(currentFragment.remaining(), destination.remaining());
+        final int originalBufferLimit = currentFragment.limit();
+        currentFragment.limit(currentFragment.position() + limit);
+        destination.put(currentFragment);
+        // Need to reset buffer limit so we can read from the same buffer again in future
+        currentFragment.limit(originalBufferLimit);
+        position = localPosition + limit;
+        return limit;
+    }
+
+    private int fragmentIndexForByteNumber(final long byteNumber) {
+        return Math.toIntExact(byteNumber / fragmentSize);
+    }
+
+    private CompletableFuture<ByteBuffer> computeFragmentFuture(final int fragmentIndex) {
+        final long readFrom = (long) fragmentIndex * fragmentSize;
+        final long readTo = Math.min(readFrom + fragmentSize, size) - 1;
+        final String range = "bytes=" + readFrom + "-" + readTo;
```

Comment:
https://www.rfc-editor.org/rfc/rfc9110.html#name-content-range suggests that if you know the size of the underlying object, you should set it.
`"bytes=" + readFrom + "-" + readTo + "/" + size`

Reply:
I found the didn't find the suggestion in the RFC, can you please help me locate it?

I think bytes=start-end is more universal. When I looked at the header, it says `Accept-Ranges -> bytes`. So I don't know how exactly to check if it supports content range requests.
I also couldn't find any way to fetch content-ranges using the java client. When I send `"bytes=" + readFrom + "-" + readTo + "/" + size` in the regular range method `builder.bucket(bucket).key(key).range(range)`, it leads to a segfault.

Reply:
The link I sent says:
> For byte ranges, a sender SHOULD indicate the complete length of the representation from which the range has been extracted, unless the complete length is unknown or difficult to determine.

Reply:
Ah, I think you are right. And I think Content-Range is a *response* not a request. Sorry.